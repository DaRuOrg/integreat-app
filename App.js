import App from './src/modules/app/components/App'

export default App
