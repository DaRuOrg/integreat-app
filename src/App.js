import React from 'react'
import { Provider } from 'react-redux'
import { Route, Router, Switch } from 'react-router-dom'

import PropTypes from 'prop-types'

import store from './store'

import LandingPage from './LandingPage'
import LocationPage from './LocationPage'
import SearchPage from './LocationPage/SearchPage'
import Hierarchy from './LocationPage/Hierarchy'
import ErrorPage from './ErrorPage'
import { I18nextProvider } from 'react-i18next'
import i18n from './i18n/i18n'

class App extends React.Component {
  static propTypes = {
    history: PropTypes.any.isRequired
  }

  render () {
    return (
      <I18nextProvider i18n={ i18n }>
        <Provider store={store}>
          <Router history={this.props.history}>
            <Switch>
              {/* The root page */}
              <Route path="/" exact component={LandingPage}/>
              {/* The location page */}
              <Route path="/location/:location/search" exact component={SearchPage}/>
              <Route path="/location/:location/:path*" render={props => {
                let path = props.match.params.path
                return <LocationPage {...props} hierarchy={new Hierarchy(path)}/>
              }}/>
              {/* The error page */}
              <Route component={ErrorPage}/>
            </Switch>
          </Router>
        </Provider>
      </I18nextProvider>
    )
  }
}

export default App
