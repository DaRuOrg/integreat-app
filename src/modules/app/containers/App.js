import React from 'react'
import PropTypes from 'prop-types'

import { Provider } from 'react-redux'
import { I18nextProvider } from 'react-i18next'

import LandingPage from 'routes/landing/containers/LandingPage'
import CategoriesPage from 'routes/categories/containers/CategoriesPage'
import SearchPage from 'routes/search/containers/SearchPage'
import DisclaimerPage from 'routes/disclaimer/containers/DisclaimerPage'
import EventsPage from 'routes/events/containers/EventsPage'
import PdfFetcherPage from 'routes/pdf-fetcher/containers/PdfFetcherPage'
import MainDisclaimerPage from 'routes/main-disclaimer/containers/MainDisclaimerPage'
import RedirectPage from 'routes/redirect/containers/RedirectPage'
import { Fragment } from 'redux-little-router'
import Layout from 'modules/app/components/Layout'
import Store from 'Store'
import I18n from 'I18n'
import Footer from './Footer'
import AutoHeader from './AutoHeader'

class App extends React.Component {
  static propTypes = {
    store: PropTypes.instanceOf(Store),
    i18n: PropTypes.instanceOf(I18n)
  }

  render () {
    const LandingLayout = ({children}) => <Layout footer={<Footer />}>{children}</Layout>
    const RichLayout = ({children}) => <Layout footer={<Footer />} header={<AutoHeader />}>{children}</Layout>

    return <I18nextProvider i18n={this.props.i18n.i18next}>
      <Provider store={this.props.store.redux}>
        {/*
          * For routes inside a <React.Fragment /> the priority decreases with each element
          * So /disclaimer has higher priority than /:language -> '/disclaimer' resolves to /disclaimer
          */}
        <Fragment forRoute='/'>
          {/* Routes */}
          <React.Fragment>

            {/* Matches two or more arguments like /augsburg/de */}
            <Fragment forRoute='/:location/:language'>
              <React.Fragment>
                {/* Matches /augsburg/de/search -> Search */}
                <Fragment forRoute='/search'>
                  <RichLayout><SearchPage /></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/disclaimer -> Disclaimer */}
                <Fragment forRoute='/disclaimer'>
                  <RichLayout><DisclaimerPage /></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/events* -> Events */}
                <Fragment forRoute='/events*'>
                  <RichLayout><EventsPage /></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/redirect -> Redirect */}
                <Fragment forRoute='/redirect'>
                  <RichLayout><RedirectPage /></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/fetch-pdf/* -> Redirect */}
                <Fragment forRoute='/fetch-pdf/*'>
                  <Layout><PdfFetcherPage /></Layout>
                </Fragment>
                {/* Matches /augsburg/de/* -> Content */}
                <Fragment forNoMatch>
                  <RichLayout><CategoriesPage /></RichLayout>
                </Fragment>
              </React.Fragment>
            </Fragment>

            {/* Matches /disclaimer */}
            <Fragment forRoute='/disclaimer'>
              <RichLayout><MainDisclaimerPage /></RichLayout>
            </Fragment>

            {/* Matches one argument like /de */}
            <Fragment forRoute='/:language'>
              <LandingLayout><LandingPage /></LandingLayout>
            </Fragment>

            {/* Matches zero arguments like / */}
            <Fragment forRoute='/'>
              <LandingLayout><LandingPage /></LandingLayout>
            </Fragment>

            {/* There are no missing routes. Covered:
              * Two or more arguments (Search/Disclaimer/Events/PdfFetcher/CategoriesPage)
              * One argument (MainDisclaimer or LandingPage with language preselection)
              * No arguments (LandingPage)
              */}

          </React.Fragment>
        </Fragment>
      </Provider>
    </I18nextProvider>
  }
}

export default App
