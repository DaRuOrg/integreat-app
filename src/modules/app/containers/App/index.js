import React from 'react'
import PropTypes from 'prop-types'

import { Provider } from 'react-redux'
import { I18nextProvider } from 'react-i18next'

import LandingPage from 'routes/landing/containers/LandingPage'
import LocationPage from 'routes/location/containers/LocationPage'
import SearchPage from 'routes/search/containers/SearchPage'
import ErrorPage from 'routes/error/containers/ErrorPage'
import DisclaimerPage from 'routes/disclaimer/containers/DisclaimerPage'
import EventsPage from 'routes/events/containers/EventsPage'
import PdfFetcherPage from 'routes/pdf-fetcher/containers/PdfFetcherPage'
import MainDisclaimerPage from 'routes/main-disclaimer/containers/MainDisclaimerPage'
import PageRedirectorPage from 'routes/redirect/containers/PageRedirectorPage'
import { Fragment } from 'redux-little-router'
import RichLayout from 'modules/app/containers/RichLayout'
import Layout from 'modules/app/components/Layout'
import Store from 'modules/app/Store'

class App extends React.Component {
  static propTypes = {
    store: PropTypes.instanceOf(Store),
    i18n: PropTypes.instanceOf(Store)
  }

  render () {
    return <I18nextProvider i18n={this.props.i18n.i18next}>
      <Provider store={this.props.store.redux}>
        {/*
         For routes inside a <div/> the priority decreases with each element
         So /disclaimer has higher priority than /:language -> '/disclaimer' resolves to /disclaimer
      */}
        <Fragment forRoute="/">
          {/* Routes */}
          <div>
            {/* Matches /disclaimer */}
            <Fragment forRoute="/disclaimer">
              <RichLayout><MainDisclaimerPage/></RichLayout>
            </Fragment>
            {/* Matches / */}
            <Fragment forRoute="/">
              <Layout><LandingPage/></Layout>
            </Fragment>

            {/* Matches /augsburg/de */}
            <Fragment forRoute="/:location/:language">
              <div>
                {/* Matches /augsburg/de/search -> Search */}
                <Fragment forRoute="/search">
                  <RichLayout><SearchPage/></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/disclaimer -> Disclaimer */}
                <Fragment forRoute="/disclaimer">
                  <RichLayout><DisclaimerPage/></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/events -> Events */}
                <Fragment forRoute="/events*">
                  <RichLayout><EventsPage/></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/redirect -> Redirect */}
                <Fragment forRoute="/redirect">
                  <RichLayout><PageRedirectorPage/></RichLayout>
                </Fragment>
                {/* Matches /augsburg/de/fetch-pdf/* -> Redirect */}
                <Fragment forRoute="/fetch-pdf/*">
                  <Layout><PdfFetcherPage/></Layout>
                </Fragment>
                {/* Matches /augsburg/de/* -> Location */}
                <Fragment forRoute="*">
                  <RichLayout><LocationPage/></RichLayout>
                </Fragment>
              </div>
            </Fragment>

            {/* Matches /de */}
            <Fragment forRoute="/:language">
              <Layout><LandingPage/></Layout>
            </Fragment>

            <Fragment forNoRoute>
              <RichLayout><ErrorPage/></RichLayout>
            </Fragment>
          </div>
        </Fragment>
      </Provider>
    </I18nextProvider>
  }
}

export default App
