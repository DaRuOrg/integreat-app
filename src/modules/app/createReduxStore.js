// @flow

import type { Store } from 'redux'
import { applyMiddleware, combineReducers, compose, createStore } from 'redux'
import { createLogger } from 'redux-logger'

import { languageReducer, uiDirectionReducer } from '../i18n/reducers'
import toggleDarkModeReducer from '../theme/reducers'
import Payload from '../endpoint/Payload'

export type ActionType<T> = { type: string, payload: Payload<T> }

// todo: Change type to correct State type,
// https://blog.callstack.io/type-checking-react-and-redux-thunk-with-flow-part-2-206ce5f6e705
const createReduxStore = (initialState: {} = {}): Store<any, any> => {
  /**
   * The middlewares of this app, add additional middlewares here
   */
  const middlewares = []

  if (__DEV__) {
    middlewares.push(createLogger()) // Logs all state changes in console
  }

  const rootReducer = combineReducers({
    uiDirection: uiDirectionReducer,
    language: languageReducer,
    darkMode: toggleDarkModeReducer
  })

  const enhancers = compose(applyMiddleware(...middlewares))

  return createStore(rootReducer, initialState, enhancers)
}

export default createReduxStore
