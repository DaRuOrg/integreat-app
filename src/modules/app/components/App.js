// @flow

import * as React from 'react'
import PlatformContext from '../../platform/PlatformContext'
import Platform from '../../platform/Platform'

import { Provider } from 'react-redux'
import I18nProvider from 'modules/i18n/containers/I18nProvider'
import createReduxStore from '../createReduxStore'
import CustomThemeProvider from '../../theme/containers/CustomThemeProvider'
import Navigator from './Navigator'
import IOSSafeAreaView from 'modules/platform/components/IOSSafeAreaView'
import AndroidStatusBarContainer from '../../platform/containers/AndroidStatusBarContainer'

class App extends React.Component<{}, { waitingForStore: boolean }> {
  platform: Platform

  store: any

  constructor () {
    super()
    this.state = {waitingForStore: true}
    this.store = createReduxStore(() => { this.setState({waitingForStore: false}) })
    this.platform = new Platform()
  }

  render () {
    if (this.state.waitingForStore) {
      return null
    }

    return (
      <Provider store={this.store}>
          <I18nProvider>
            <CustomThemeProvider>
              <PlatformContext.Provider value={this.platform}>
                <AndroidStatusBarContainer />
                <IOSSafeAreaView>
                  <Navigator />
                </IOSSafeAreaView>
              </PlatformContext.Provider>
            </CustomThemeProvider>
          </I18nProvider>
      </Provider>
    )
  }
}

export default App
