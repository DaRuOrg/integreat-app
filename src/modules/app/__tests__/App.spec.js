import { shallow } from 'enzyme'
import App from '../containers/App'
import React from 'react'
import createReduxStore from '../createReduxStore'
import createHistory from '../createHistory'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'

const mockStore = configureMockStore()
const mockedStore = mockStore()

jest.mock('../createReduxStore', () => jest.fn().mockImplementation(() => mockedStore))

describe('App', () => {
  test('should match snapshot', () => {
    const component = shallow(<App />)
    expect(component).toMatchSnapshot()
  })

  test('should create correct store and pass it to Provider', () => {
    const app = shallow(<App />)

    expect(createReduxStore).toHaveBeenCalledWith(createHistory)
    expect(app.find(Provider).prop('store')).toEqual(mockedStore)
  })
})
