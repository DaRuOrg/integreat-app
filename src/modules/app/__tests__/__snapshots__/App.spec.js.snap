// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "matchRoute": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Connect(I18nProvider)>
    <RouterFragment />
  </Connect(I18nProvider)>
</Provider>
`;
